airbyte-cdk/python/unit_tests/base_python/test_abstract_source.py

```diff
@@ -63,34 +114,173 @@ def test_discover(mocker):
+            msg.record.emitted_at = GLOBAL_EMITTED_AT
+    return messages
 
 
-def test_valid_full_refresh_read_with_slices():
-    pass
+def test_valid_full_refresh_read_no_slices(logger, mocker):
+    stream_output = [{"k1": "v1"}, {"k2": "v2"}]
+    s1 = MockStream([({"sync_mode": SyncMode.full_refresh}, stream_output)], name="s1")
+    s2 = MockStream([({"sync_mode": SyncMode.full_refresh}, stream_output)], name="s2")
 
+    mocker.patch.object(MockStream, "get_json_schema", return_value={})
+
+    src = MockSource(streams=[s1, s2])
+    catalog = ConfiguredAirbyteCatalog(
+        streams=[_configured_stream(s1, SyncMode.full_refresh), _configured_stream(s2, SyncMode.full_refresh)]
+    )
 
-def test_valid_incremental_read_with_record_interval():
-    pass
+    expected = _as_records("s1", stream_output) + _as_records("s2", stream_output)
+    messages = _fix_emitted_at(list(src.read(logger, {}, catalog)))
 
+    assert expected == messages
 
-def test_valid_incremental_read_with_no_interval():
-    pass
 
+def test_valid_full_refresh_read_with_slices(mocker, logger):
+    stream_output = [{"k1": "v1"}, {"k2": "v2"}]
+    slices = [{"1": "1"}, {"2": "2"}]
+    s1 = MockStream([({"sync_mode": SyncMode.full_refresh, "stream_slice": s}, stream_output) for s in slices], name="s1")
+    s2 = MockStream([({"sync_mode": SyncMode.full_refresh, "stream_slice": s}, stream_output) for s in slices], name="s2")
+
+    mocker.patch.object(MockStream, "get_json_schema", return_value={})
+    mocker.patch.object(MockStream, "stream_slices", return_value=slices)
+
+    src = MockSource(streams=[s1, s2])
+    catalog = ConfiguredAirbyteCatalog(
+        streams=[_configured_stream(s1, SyncMode.full_refresh), _configured_stream(s2, SyncMode.full_refresh)]
+    )
+
+    expected = _as_records("s1", stream_output) * 2 + _as_records("s2", stream_output) * 2
+    messages = _fix_emitted_at(list(src.read(logger, {}, catalog)))
 
-def test_valid_incremental_read_with_slices():
-    pass
+    assert expected == messages
+
+
+def _state(state_data: Dict[str, Any]):
+    return AirbyteMessage(type=Type.STATE, state=AirbyteStateMessage(data=state_data))
+
+
+def test_valid_incremental_read_with_checkpoint_interval(mocker, logger):
+    stream_output = [{"k1": "v1"}, {"k2": "v2"}]
+    s1 = MockStream([({"sync_mode": SyncMode.incremental, "stream_state": {}}, stream_output)], name="s1")
+    s2 = MockStream([({"sync_mode": SyncMode.incremental, "stream_state": {}}, stream_output)], name="s2")
+    state = {"cursor": "value"}
+    mocker.patch.object(MockStream, "get_updated_state", return_value=state)
+    mocker.patch.object(MockStream, "supports_incremental", return_value=True)
+    mocker.patch.object(MockStream, "get_json_schema", return_value={})
+    # Tell the source to output one state message per record
+    mocker.patch.object(MockStream, "state_checkpoint_interval", new_callable=mocker.PropertyMock, return_value=1)
```

Comment:
@sherifnada this `PropertyMock` was what was missing.

Reply:
thanks boss